base/broadcast.jl

```diff
@@ -219,6 +219,12 @@ argtype(bc::Broadcasted) = argtype(typeof(bc))
 _eachindex(t::Tuple{Any}) = t[1]
 _eachindex(t::Tuple) = CartesianIndices(t)
 
+Base.IndexStyle(bc::Broadcasted) = IndexStyle(typeof(bc))
+Base.IndexStyle(::Type{<:Broadcasted{<:Any,Tuple{Axis}}}) where {Axis} = IndexStyle(Axis)
+Base.IndexStyle(::Type{<:Broadcasted{<:Any}}) = IndexCartesian()
```

Comment:
It turned out my previous approach (combine `IndexStyle` of each argument) was wrong since it didn't work with, e.g., `broadcasted(+, zeros(5), zeros(1, 4))`.  Each argument can be `IndexLinear` even though the broadcasted indexing is not (that's the main job of broadcasted).

So now the new approach is to return an equivalent of `IndexStyle(bc.axes[1])` if `length(bc.axes) == 1` and then return `IndexCartesian()` otherwise.  As I need `Axes` type parameter now, this implementation requires the `Broadcasted` to be `instantiate`d first.  I guess this is OK since that's the usual broadcasting pipeline.

Is this implementation fine?  Is `IndexStyle(bc.axes[1]) == IndexLinear() && length(bc.axes) == 1` the only possible `IndexLinear()` case that is detectable in a type-stable manner?  That is to say, I'm ignoring the case like `broadcasted(+, zeros(1, 4), zeros(1, 4))` as I need to look at the value to check that it supports linear indexing.

Reply:
This doesn't really require the `Broadcasted` to be `instantiate`d to get an `IndexStyle` — it just means that if it's not instantiated that it falls back to a cartesian implementation. All `Broadcasted`s should support cartesian indexing; only one-dimensional ones will allow straight integers.

Reply:
Yes, that's much more accurate way to put it.  I wanted to say that "for the broadcasted to be dispatched to the optimized method", it has to be `instantiate`d first.